itest/lnd_sendonion_test.go

```diff
@@ -142,4 +151,181 @@ func testSendOnion(ht *lntest.HarnessTest) {
+		alice, bob, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(alice, chanPointAlice)
+
+	// We'll create Dave and establish a channel to Alice. Dave will be
+	// running an older node that requires the legacy onion payload.
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, dave)
+	chanPointBob := ht.OpenChannel(
+		bob, carol, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(bob, chanPointBob)
+
+	// Next, we'll create Carol and establish a channel to from her to Dave.
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, carol)
+	chanPointCarol := ht.OpenChannel(
+		carol, dave, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(carol, chanPointCarol)
+
+	// Make sure Alice knows the channel between Bob and Carol.
+	ht.AssertTopologyChannelOpen(alice, chanPointBob)
+	ht.AssertTopologyChannelOpen(alice, chanPointCarol)
+
+	const paymentAmt = 10000
+
+	// Query for routes to pay from Alice to Dave.
+	routesReq := &lnrpc.QueryRoutesRequest{
+		PubKey: dave.PubKeyStr,
+		Amt:    paymentAmt,
+	}
+	routes := alice.RPC.QueryRoutes(routesReq)
+	route := routes.Routes[0]
+
+	finalHop := route.Hops[len(route.Hops)-1]
+	finalHop.MppRecord = &lnrpc.MPPRecord{
+		PaymentAddr:  ht.Random32Bytes(),
+		TotalAmtMsat: int64(lnwire.NewMSatFromSatoshis(paymentAmt)),
+	}
+
+	ht.Logf("Found route from Alice to Dave: %+v", route)
+
+	// Build the onion to use for our payment.
+	paymentHash := ht.Random32Bytes()
+	onionReq := &routerrpc.BuildOnionRequest{
+		Route:       route,
+		PaymentHash: paymentHash,
+	}
+	onionResp := alice.RPC.BuildOnion(onionReq)
+	ht.Logf("Constructed onion: %+v w/ key: %x", onionResp.OnionBlob,
+		onionResp.SessionKey)
+
+	// Dispatch a payment via SendOnion.
+	firstHop := bob.PubKey
+	sendReq := &routerrpc.SendOnionRequest{
+		FirstHopPubkey: firstHop[:],
+		Amount:         route.TotalAmtMsat,
+		Timelock:       route.TotalTimeLock,
+		PaymentHash:    paymentHash,
+		OnionBlob:      onionResp.OnionBlob,
+		AttemptId:      1,
+	}
+	ht.Logf("Sending onion w/ amt=%d (msat) to %x",
+		sendReq.Amount, firstHop)
+
+	resp := alice.RPC.SendOnion(sendReq)
+	ht.Logf("SendOnion response: %+v", resp)
+
+	serverErrorStr := ""
+	clientErrorStr := ""
+
+	// Track the payment providing all necessary information to delegate
+	// error decryption to the server.
+	//
+	// NOTE(calvin): We expect this to fail as Dave is not expecting payment.
+	ctxt, _ := context.WithTimeout(context.Background(), defaultTimeout)
+	trackReq := &routerrpc.TrackOnionRequest{
+		AttemptId:   1,
+		PaymentHash: paymentHash,
+		SessionKey:  onionResp.SessionKey,
+		HopPubkeys:  onionResp.HopPubkeys,
+	}
+	trackResp, clearErr := alice.RPC.Router.TrackOnion(ctxt, trackReq)
+	if clearErr != nil {
+		ht.Logf("Encountered error while tracking onion: %v", clearErr)
+	}
+	ht.Logf("Tracked payment via onion: %+v", trackResp)
+	serverErrorStr = clearErr.Error()
+
+	// Now we'll track the same payment attempt, but we'll specify that
```

Comment:
Trying to figure out the error handling across RPC boundary between server and client. Seems we could either use the [gRPC **_status_** package](https://pkg.go.dev/google.golang.org/grpc/internal/status) or the **_(Send/Track)OnionResponse_** proto message itself to transport "rich error detail" (code + message) between server and client. 

Depending on the RPC client (eg: lightning proxy with central instantiation of _**ChannelRouter**_, self-sovereign client with hosted node operator service acting as oblivious sender, etc.), these errors would be handled differently.